x/cork/keeper/keeper.go

```diff
@@ -111,6 +113,141 @@ func (k Keeper) IterateCorks(ctx sdk.Context, handler func(val sdk.ValAddress, c
 	}
 }
 
+/////////////////////
+// Scheduled Corks //
+/////////////////////
+
+func (k Keeper) SetScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, cork types.Cork) {
+	bz := k.cdc.MustMarshal(&cork)
+	ctx.KVStore(k.storeKey).Set(types.GetScheduledCorkKey(val, blockHeight, common.HexToAddress(cork.TargetContractAddress)), bz)
+}
+
+// GetScheduledCork gets the scheduled cork for a given validator
+// CONTRACT: must provide the validator address here not the delegate address
+func (k Keeper) GetScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, contract common.Address) (types.Cork, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetScheduledCorkKey(val, blockHeight, contract))
+	if len(bz) == 0 {
+		return types.Cork{}, false
+	}
+
+	var cork types.Cork
+	k.cdc.MustUnmarshal(bz, &cork)
+	return cork, true
+}
+
+// DeleteScheduledCork deletes the scheduled cork for a given validator
+// CONTRACT: must provide the validator address here not the delegate address
+func (k Keeper) DeleteScheduledCork(ctx sdk.Context, val sdk.ValAddress, blockHeight uint64, contract common.Address) {
+	ctx.KVStore(k.storeKey).Delete(types.GetScheduledCorkKey(val, blockHeight, contract))
+}
+
+// IterateScheduledCorks iterates over all scheduled corks in the store
+func (k Keeper) IterateScheduledCorks(ctx sdk.Context, cb func(val sdk.ValAddress, blockHeight uint64, cel common.Address, cork types.Cork) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, []byte{types.ScheduledCorkForAddressKey})
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		var cork types.Cork
+		keyPair := bytes.NewBuffer(bytes.TrimPrefix(iter.Key(), []byte{types.ScheduledCorkForAddressKey}))
+		blockHeight := binary.BigEndian.Uint64(keyPair.Next(8))
+		val := sdk.ValAddress(keyPair.Next(20))
+		cel := common.BytesToAddress(keyPair.Bytes())
+
+		k.cdc.MustUnmarshal(iter.Value(), &cork)
+		if cb(val, blockHeight, cel, cork) {
+			break
+		}
+	}
+}
+
+///////////////////////////
+// ScheduledBlockHeights //
```

Comment:
Isn't the list of scheduled block heights implicit in the first level of the `ScheduledCork` keyspace since it's effectively the index? What do we gain by having a separate keyspace?

Reply:
Wouldn't we need to iterate all of the scheduled corks in order to read all of the block heights then? I thought you wanted a secondary keyspace?